test/src/unittests/rhythm/test_loopbpmconfidence.py

```diff
@@ -0,0 +1,213 @@
+        self.assertAlmostEqual(expectedConfidence, confidence, 0.01) 
+        
+        bpmEstimate = 70
+        expectedConfidence = 0.50 # rounded from measured 0.5011640191078186
+        confidence = LoopBpmConfidence()(audio[0:len90], bpmEstimate)                     
+        self.assertAlmostEqual(expectedConfidence, confidence, 0.01)
+        
+        # Truncated Audios 90% and 1% at beginning
+        bpmEstimate = 125
+        expectedConfidence = 0.92 # rounded from measured 0.9199735522270203
+        confidence = LoopBpmConfidence()(audio[len01:len90], bpmEstimate)                         
+        self.assertAlmostEqual(expectedConfidence, confidence, 0.01) 
+
+        bpmEstimate = 120
+        expectedConfidence =  0.36 # rounded from measured 0.3631746172904968
+        confidence = LoopBpmConfidence()(audio[len01:len90], bpmEstimate)           
+        self.assertAlmostEqual(expectedConfidence, confidence, 0.01) 
+        
+        bpmEstimate = 70 
+        expectedConfidence =  0.79 # rounded from measured 0.7951852083206177
+        confidence = LoopBpmConfidence()(audio[len01:len90], bpmEstimate)
+        self.assertAlmostEqual(expectedConfidence, confidence, 0.01) 
+
+
+    """
+    The confidence returned is based on comparing the duration of the audio signal with 
+    multiples of the BPM estimate. 
+    The algorithm considers edge cases in which the signal includes non-musical 
+    silence at the start or end of the audio signal (or both) and returns the confidence for the best fit.
+    For more details refer to paper: https://repositori.upf.edu/handle/10230/33113
+
+    Additional Tests were made with following loop
+    "458538__flowerpunkchip__mellifluousanonymous-120bpm-loop-youareevil.mp3'"
+    https://freesound.org/people/Flowerpunkchip/sounds/458538/
+
+    Front end slience Confidence is  0.9915646314620972
+    Back end silence Confidence  is  0.9915646314620972
+    Silence Both end confidence is  0.9915646314620972    
+    """
+    def testSilentEdge(self):
+        audio = MonoLoader(filename=join(testdata.audio_dir, 'recorded', 'techno_loop.wav'))() 
+        bpmEstimate = 125
+        confidence = LoopBpmConfidence()(audio, bpmEstimate)              
+        lenSilence = 30000 # N.B The beat period is 21168 samples for 125 bpm @ 44.1k samp. rate
+        silentAudio = zeros(lenSilence)
+        benchmarkConfidence = 0.96 # This figure was arrived at emperically from the min. confidence observed with test runs 
+
+        # Test addition of non-musical silence before the loop starts
+        # The length is not a beat period,
+        # Nonetheless, we can stillreliably estimate the starting point because it is a hard transient.
+        signal1  = numpy.append(silentAudio, audio)
+        confidence = LoopBpmConfidence()(signal1, bpmEstimate)      
+        self.assertGreater(confidence, benchmarkConfidence)
+
+    def testExactAudioLengthMatch(self):
```

Comment:
I would add a test on zeros(beatPeriod * 4) which should also give confidence value of 1.